src/bip340.rs

```diff
@@ -0,0 +1,524 @@
+    pub fn as_mut_ptr(&mut self) -> *mut ffi::XOnlyPublicKey {
+        &mut self.0 as *mut _
+    }
+
+    /// Creates a new BIP340 public key from a secret key.
+    #[inline]
+    pub fn from_secret_key<C: Signing>(secp: &Secp256k1<C>, sk: &SecretKey) -> PublicKey {
+        let mut keypair = ffi::KeyPair::new();
+        let mut xonly_pk = ffi::XOnlyPublicKey::new();
+        let mut pk_parity = 0;
+        unsafe {
+            let mut ret = ffi::secp256k1_keypair_create(secp.ctx, &mut keypair, sk.as_c_ptr());
+            debug_assert_eq!(ret, 1);
+            ret =
+                ffi::secp256k1_keypair_xonly_pub(secp.ctx, &mut xonly_pk, &mut pk_parity, &keypair);
+            debug_assert_eq!(ret, 1);
+        }
+        PublicKey(xonly_pk)
+    }
+
+    /// Creates a BIP340 public key directly from a slice
+    #[inline]
+    pub fn from_slice(data: &[u8]) -> Result<PublicKey, Error> {
+        if data.is_empty() || data.len() != constants::BIP340_PUBLIC_KEY_SIZE {
+            return Err(InvalidPublicKey);
+        }
+
+        let mut pk = ffi::XOnlyPublicKey::new();
+        unsafe {
+            if ffi::secp256k1_xonly_pubkey_parse(
+                ffi::secp256k1_context_no_precomp,
+                &mut pk,
+                data.as_c_ptr(),
+            ) == 1
+            {
+                Ok(PublicKey(pk))
+            } else {
+                Err(InvalidPublicKey)
+            }
+        }
+    }
+
+    #[inline]
+    /// Serialize the key as a byte-encoded pair of values. In compressed form
+    /// the y-coordinate is represented by only a single bit, as x determines
+    /// it up to one bit.
+    pub fn serialize(&self) -> [u8; constants::BIP340_PUBLIC_KEY_SIZE] {
+        let mut ret = [0; constants::BIP340_PUBLIC_KEY_SIZE];
+
+        unsafe {
+            let err = ffi::secp256k1_xonly_pubkey_serialize(
+                ffi::secp256k1_context_no_precomp,
+                ret.as_mut_c_ptr(),
+                self.as_c_ptr(),
+            );
+            debug_assert_eq!(err, 1);
+        }
+        ret
+    }
+}
+
+impl CPtr for PublicKey {
+    type Target = ffi::XOnlyPublicKey;
+    fn as_c_ptr(&self) -> *const Self::Target {
+        self.as_ptr()
+    }
+
+    fn as_mut_c_ptr(&mut self) -> *mut Self::Target {
+        self.as_mut_ptr()
+    }
+}
+
+/// Creates a new BIP340 public key from a FFI x-only public key
+impl From<ffi::XOnlyPublicKey> for PublicKey {
+    #[inline]
+    fn from(pk: ffi::XOnlyPublicKey) -> PublicKey {
+        PublicKey(pk)
+    }
+}
+
+serde_impl_from_slice!(PublicKey);
+
+impl<C: Signing> Secp256k1<C> {
+    /// Create a bip340 signature using OsRng to generate the auxiliary random
+    /// data. Requires compilation with "rand-std" feature.
+    #[cfg(any(test, feature = "rand-std"))]
+    pub fn bip340_sign(&self, msg: &Message, sk: &SecretKey) -> Result<Signature, Error> {
+        let mut rng = OsRng::new().expect("OsRng");
+        self.bip340_sign_with_rng(msg, sk, &mut rng)
+    }
```

Comment:
This should call 
```rust
                ffi::secp256k1_schnorrsig_sign(
                    self.ctx,
                    sig.as_mut_c_ptr(),
                    msg.as_c_ptr(),
                    &keypair,
                    ffi::secp256k1_nonce_function_bip340,
                    ptr::null(),
                )
```

Reply:
Oh ok now I understand what you want. But is it a good idea to have the most obvious function name `bip340_sign` for this then? My understanding of bip340 is that using random auxiliary data should be encouraged, but here I'm afraid that this will become the default choice for users of the library. Maybe then it should be:
```rust
bip340_sign<R: Rng + CryptoRng>(
        &self,
        msg: &Message,
        sk: &SecretKey,
        rng: &mut R,
    )
```
```rust
bip340_sign_with_aux_rand(
        &self,
        msg: &Message,
        sk: &SecretKey,
        aux_rand: &[u8; 32],
    )
```
```rust 
bip340_sign_no_aux_rand(
        &self,
        msg: &Message,
        sk: &SecretKey
    )
```

What do you think?

Reply:
I'm ok with these, just the 2nd one I'm ~0 on it (not against but not for)

Reply:
Updated as mentioned above (with some refactor to avoid duplicate code).

Also I skimmed through [the comments](https://github.com/bitcoin-core/secp256k1/blob/master/include/secp256k1_schnorrsig.h#L52) in the upstream repo and they mention that passing null "does not produce BIP-340 compliant signatures" so just wanted to confirm that you really want to give this option.

Reply:
Hm, I think the default `sign` function should simply pass `thread_rng` here (at least if we have `std`). I'd be fine with adding yet another function `bip340_sign_with_aux_rng`. We could maybe drop `bip340_sign_with_aux_rand` instead, I'm not sure if this is helpful in Rust. Does this make sense?

Reply:
@real-or-random You want to provide 2 separate `sign` functions depending on if `rand-std` is enabled?

Reply:
what I had in mind is something along these lines:
 - if  `rand-std` enabled: `_sign`
 - always: `_sign_with_aux_rng` and `_sign_without_aux_rand` (or `sign_without_aux_rng`)

Reply:
Okay, now that I looked at the current version, I think having `_sign_with_aux_rand` is good to have because it does not need the rand module.

So I believe we should have
 * if `rand-std` enabled: `_sign`
 * if `rand` enabled: `_sign_with_aux_rng` (this should really be `_sign_with_aux_rng` and only `_sign_with_prg` because the latter sounds like we don't use deterministic nonces...)
 * always: `_sign_with_aux_rand` and `_sign_without_aux_rand`.


But let's first agree on this before you need to go back and forth with that as well. @elichai @apoelstra Does this sound good? These are four functions. May be overkill but I think all have their reason to exist.